fix(deps): update from eth-rpc-errors to @metamask/rpc-errors (cause edition)

legobeat opened the pull request:
## **Description**

- Upgrade from obsolete `eth-rpc-errors` to `@metamask/rpc-errors`
  - This introduce handling of error causes 


See [here](https://github.com/MetaMask/rpc-errors/pull/140) for some context.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/24496?quickstart=1)

## **Related issues**

- #22871

#### Blocked by
- [x] https://github.com/MetaMask/rpc-errors/pull/158
- [x] https://github.com/MetaMask/rpc-errors/pull/144
  - [x] https://github.com/MetaMask/rpc-errors/pull/140

#### Blocking
- #22875

## **Manual testing steps**

## **Screenshots/Recordings**

### **Before**

### **After**

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

legobeat commented:
[Failing](https://app.circleci.com/pipelines/github/MetaMask/metamask-extension/80331/workflows/c3dd9f1d-9174-418a-8de2-5e8c126b36b3/jobs/2832444) `test-unit-jest-main`:

<details>

```
$ yarn test:coverage:jest

[...]

FAIL app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts
  ● metamask-notifications - markMetamaskNotificationsAsRead() › updates feature announcements and wallet notifications as read

    expect(received).toBe(expected) // Object.is equality

    Expected: 2
    Received: 1

      548 |
      549 |     // Should see 2 items in controller read state
    > 550 |     expect(controller.state.metamaskNotificationsReadList.length).toBe(2);
          |                                                                   ^
      551 |   });
      552 |
      553 |   test('should at least mark feature announcements locally if external updates fail', async () => {

      at Object.toBe (app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts:550:67)

  ● metamask-notifications - markMetamaskNotificationsAsRead() › should at least mark feature announcements locally if external updates fail

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      563 |     // This is because on-chain failed.
      564 |     // We can debate & change implementation if it makes sense to mark as read locally if external APIs fail.
    > 565 |     expect(controller.state.metamaskNotificationsReadList.length).toBe(1);
          |                                                                   ^
      566 |   });
      567 |
      568 |   function arrangeMocks(options?: { onChainMarkAsReadFails: boolean }) {

      at Object.toBe (app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts:565:67)

[...]

Summary of all failing tests
FAIL app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts
  ● metamask-notifications - markMetamaskNotificationsAsRead() › updates feature announcements and wallet notifications as read

    expect(received).toBe(expected) // Object.is equality

    Expected: 2
    Received: 1

      548 |
      549 |     // Should see 2 items in controller read state
    > 550 |     expect(controller.state.metamaskNotificationsReadList.length).toBe(2);
          |                                                                   ^
      551 |   });
      552 |
      553 |   test('should at least mark feature announcements locally if external updates fail', async () => {

      at Object.toBe (app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts:550:67)

  ● metamask-notifications - markMetamaskNotificationsAsRead() › should at least mark feature announcements locally if external updates fail

    expect(received).toBe(expected) // Object.is equality

    Expected: 1
    Received: 0

      563 |     // This is because on-chain failed.
      564 |     // We can debate & change implementation if it makes sense to mark as read locally if external APIs fail.
    > 565 |     expect(controller.state.metamaskNotificationsReadList.length).toBe(1);
          |                                                                   ^
      566 |   });
      567 |
      568 |   function arrangeMocks(options?: { onChainMarkAsReadFails: boolean }) {

      at Object.toBe (app/scripts/controllers/metamask-notifications/metamask-notifications.test.ts:565:67)


Test Suites: 1 failed, 91 passed, 92 total
Tests:       2 failed, 742 passed, 744 total
Snapshots:   54 passed, 54 total
Time:        145.21 s
Ran all test suites.
```

</details>

legobeat commented:
[Passing all CI tests](https://github.com/MetaMask/metamask-extension/pull/24496/commits/94785e4e30aae0ca004bd0274304befedb5d447c).

legobeat commented on app/scripts/lib/createMetaRPCHandler.test.js:
```diff
@@ -70,7 +70,8 @@ describe('createMetaRPCHandler', () => {
     streamTest.on('data', (data) => {
-      expect(data.error.message).toStrictEqual('foo-error');
+      expect(data.error.message).toStrictEqual('Internal JSON-RPC error.');
+      expect(data.error.data.cause.message).toStrictEqual('foo-error');
```
As the `data` here is serialized JSON (and not an instance of any error object), it should still keep the `data.cause` and not have a `cause`.

legobeat commented:
This has now been updated with latest changes from https://github.com/MetaMask/rpc-errors/pull/140

legobeat commented on shared/modules/error.test.ts:
```diff
@@ -24,9 +24,9 @@ describe('error module', () => {
 
-    it('calls loglevel.error with the parameter passed in when parameter is not an instance of Error', () => {
+    it('calls loglevel.error with string representation of parameter passed in when parameter is not an instance of Error', () => {
       logErrorWithMessage({ test: 'test' });
-      expect(log.error).toHaveBeenCalledWith({ test: 'test' });
+      expect(log.error).toHaveBeenCalledWith('[object Object]');
```
Note ^

legobeat replied:
I don't see why we want this change, actually. This is now back to previous behavior.

Gudahtt commented:
Thanks @legobeat ! Could you update this and resolve conflicts? I know this has been around a while, but I'll nudge a few teams to prioritize reviewing this as well.

legobeat committed:
update createMetaRPCHandler error-handling test

legobeat committed:
deps: eth-rpc-errors@^4.0.2 -> @metamask/rpc-errors@^6.1.0

legobeat committed:
fix(multichain): read error message from cause

fix error message handling

fix: prefer cause message as error message

lint:fix

fix: update error handling for nft ownership check

fix: handle error details in error cause

fix test

fixup cause handling

Revert "fix test"

This reverts commit e5ae79be58cf22b873ede78ac8edac08832bcaac.

fix test

fix: isErrorWithMessage return type

Refactor: Handling error.cause; add pony-cause

use getErrorMessage for logging errors

WIP: deps: @metamask/rpc-errors@^6.2.1->6.3.0

test: fix testcase

fix: consider cause instead of data.cause when parsing caught error messages

chore: clean up unused cause-handling code in error.ts

chore: update lavamoat policies

revert stringification of generic objects in logErrorWithMessage

lint:fix

legobeat committed:
update lavamoat build policy

legobeat committed:
update lavamoat runtime policies

legobeat commented:
This has now been updated:
- Uses yet-to-be-published version of `@metamask/rpc-errors`
  - https://github.com/MetaMask/rpc-errors/pull/158
  - Thereby no longer changes behavior around error messages (ie: no longer clobbers with `Internal JSON-RPC error`)
  - Therefore this should now be non-breaking

legobeat committed:
deps: @metamask/rpc-errors@^6.3.0->^7.0.0

https://github.com/MetaMask/rpc-errors/pull/158

- Update test according to new @metamask/rpc-errors behavior

This partially reverts commit aeca6d797fc16d13bcd16f8649159769207ede78.

legobeat commented:
@Gudahtt After just a little back-and-forth:

[`@metamask/rpc-errors@7.0.0`](https://github.com/MetaMask/rpc-errors/releases/7.0.0) now reverts back to the previous behavior where RPC error messages are exposed directly ass `message` instead of being replace with `Internal JSON-RPC Error`. This PR now updates the direct dependency of the package to latest.

Hopefully this should clear up any outstanding concerns/reservations for this migration - let me know otherwise!

legobeat committed:
update lavamoat runtime policies

andreahaku reviewed (dismissed):
LGTM

FrederikBolding commented on ui/ducks/send/send.js:
```diff
@@ -2702,12 +2703,13 @@ export function updateSendAsset(
             } else {
               // Any other error is unexpected and should be surfaced.
-              dispatch(displayWarning(err.message));
+              dispatch(displayWarning(err));
```
Shouldn't this be `message`?

FrederikBolding commented on ui/store/actions.ts:
```diff
@@ -227,7 +229,7 @@ export function createNewVaultAndRestore(
       })
       .catch((err) => {
-        dispatch(displayWarning(err.message));
+        dispatch(displayWarning(err));
```
Seems unintentional to pass an error object directly here?

FrederikBolding commented on ui/store/actions.ts:
```diff
@@ -374,7 +376,7 @@ export function resetAccount(): ThunkAction<
         if (err) {
           if (isErrorWithMessage(err)) {
-            dispatch(displayWarning(err.message));
+            dispatch(displayWarning(err));
```
Same applies here

FrederikBolding commented on ui/pages/unlock-page/unlock-page.component.js:
```diff
@@ -95,7 +95,7 @@ export default class UnlockPage extends Component {
       this.failed_attempts += 1;
 
-      if (message === 'Incorrect password') {
+      if (message.startsWith('Incorrect password')) {
```
Curious why this is necessary now

legobeat replied:
Hm, this may not be relevant anymore but IIRC at some point things got tagged on at the end of an error message. Do you prefer it reverted to keep changes down or fine to keep as a defensive measure?

FrederikBolding replied:
I would prefer it reverted if it isn't needed, just to reduce the diff

legobeat committed:
chore: revert redundant change

Gudahtt commented:
Force-merged to override `all-jobs-passed` required status check, which is failing due to our SonarCloud checknot yet supporting forks.
